File: Sources/Deadlyne/Core/Achievements.swift
import Foundation

/// A milestone badge. Artwork is `Badges/<id>.png` (see Resources/Badges/README.md); until it
/// exists, Deadlyne draws a placeholder medallion.
struct Badge: Equatable {
    enum Track: String, CaseIterable {
        case photos, shoots

        var title: String { self == .photos ? "Photos Ingested" : "Shoots" }
        func unit(_ n: Int) -> String {
            switch self {
            case .photos: return n == 1 ? "photo" : "photos"
            case .shoots: return n == 1 ? "shoot" : "shoots"
            }
        }
    }

    let id: String
    let track: Track
    let threshold: Int
    let name: String
    /// 0 bronze, 1 silver, 2 gold, 3 elite — only the placeholder art uses it.
    let tier: Int

    /// "100", "2.5K", "10K", "1M" — for the placeholder medallion.
    var shortThreshold: String {
        if threshold >= 1_000_000 { return "\(threshold / 1_000_000)M" }
        if threshold >= 1_000 {
            return threshold % 1_000 == 0 ? "\(threshold / 1_000)K" : String(format: "%.1fK", Double(threshold) / 1_000)
        }
        return "\(threshold)"
    }
}

/// Every badge, in order. Rename them or change thresholds here — the id is the artwork file name,
/// so keep ids stable once artwork exists (earned badges are stored by id).
enum Badges {
    static let photos: [Badge] = {
        let ladder: [(Int, String)] = [
            (100, "Warm-Up"), (500, "Kickoff"), (1_000, "First Thousand"), (2_500, "Game Day"),
            (5_000, "Starter"), (10_000, "Varsity"), (25_000, "All-Conference"), (50_000, "All-State"),
            (100_000, "Hall of Fame"), (250_000, "Legend"), (500_000, "Dynasty"), (1_000_000, "The Million"),
        ]
        return ladder.enumerated().map { i, step in
            Badge(id: "photos-\(step.0)", track: .photos, threshold: step.0, name: step.1, tier: min(3, i / 3))
        }
    }()

    static let shoots: [Badge] = {
        let ladder: [(Int, String, Int)] = [
            (1, "First Shoot", 0), (10, "Double Digits", 0), (25, "Season Pass", 1),
            (50, "Road Warrior", 1), (100, "Centurion", 2), (250, "Iron Lens", 3),
        ]
        return ladder.map { Badge(id: "shoots-\($0.0)", track: .shoots, threshold: $0.0, name: $0.1, tier: $0.2) }
    }()

    static let all = photos + shoots
}

struct AchievementState: Codable {
    /// Lifetime photos brought into Deadlyne. Never goes down, even if photos are deleted later.
    var photos = 0
    var shoots = 0
    /// Photos already counted per folder (a high-water mark), so reopening a shoot or
    /// re-ingesting the same card never counts a photo twice.
    var folders: [String: Int] = [:]
    var earned: [String: Date] = [:]
    /// Photos and new shoots per calendar month, keyed "2026-09".
    var months: [String: MonthTally]?

    struct MonthTally: Codable {
        var photos = 0
        var shoots = 0
    }

    func value(_ track: Badge.Track) -> Int { track == .photos ? photos : shoots }

    var thisMonth: MonthTally { months?[Self.monthKey(Date())] ?? MonthTally() }

    static func monthKey(_ date: Date) -> String {
        let c = Calendar.current.dateComponents([.year, .month], from: date)
        return String(format: "%04d-%02d", c.year ?? 0, c.month ?? 0)
    }

    fileprivate mutating func addToMonth(photos: Int, shoots: Int) {
        let key = Self.monthKey(Date())
        var t = months?[key] ?? MonthTally()
        t.photos += photos
        t.shoots += shoots
        months = (months ?? [:]).merging([key: t]) { $1 }
    }
}

/// Counts photos as they come into Deadlyne and awards milestone badges. Main thread only.
///
/// A photo counts when it's ingested from a card, or when a folder containing it is opened for
/// the first time. A RAW+JPG pair is one photo.
enum Achievements {
    static let didChange = Notification.Name("DeadlyneAchievementsDidChange")
    /// Posted with the newly earned `[Badge]` as the object.
    static let didUnlock = Notification.Name("DeadlyneBadgesUnlocked")
    private static let key = "achievements"
    private static var cached: AchievementState?

    static var state: AchievementState {
        if let cached { return cached }
        if let data = UserDefaults.standard.data(forKey: key),
           var s = try? JSONDecoder().decode(AchievementState.self, from: data) {
            if s.months == nil {
                // Saved before monthly tallies existed: credit the totals to the month counting began.
                let start = s.earned.values.min() ?? Date()
                s.months = [AchievementState.monthKey(start): .init(photos: s.photos, shoots: s.shoots)]
                save(s)
            }
            cached = s
            return s
        }
        // First run with achievements: credit the shoots already opened, quietly.
        var s = AchievementState()
        for shoot in RecentShoots.all {
            guard let n = shoot.photoCount, n > 0 else { continue }
            s.folders[shoot.path] = n
            s.photos += n
            s.shoots += 1
        }
        _ = award(&s)
        save(s)
        return s
    }

    static func isEarned(_ b: Badge) -> Bool { state.earned[b.id] != nil }

    /// The next badge to earn on a track, or nil when all are earned.
    static func next(_ track: Badge.Track) -> Badge? {
        let value = state.value(track)
        return Badges.all.first { $0.track == track && value < $0.threshold }
    }

    /// Earned badges, oldest first.
    static var earned: [Badge] {
        let s = state
        return Badges.all.filter { s.earned[$0.id] != nil }
            .sorted { (s.earned[$0.id]!, $0.threshold) < (s.earned[$1.id]!, $1.threshold) }
    }

    /// A folder was opened and holds `count` photos. Only photos beyond what was already
    /// counted for this folder are added.
    static func recordFolder(_ url: URL, photos count: Int) {
        let key = url.standardizedFileURL.path
        let old = state.folders[key] ?? 0
        guard count > old else { return }
        mutate { s in
            let isNew = s.folders[key] == nil
            if isNew { s.shoots += 1 }
            s.folders[key] = count
            s.photos += count - old
            s.addToMonth(photos: count - old, shoots: isNew ? 1 : 0)
        }
    }

    /// Photos copied by an ingest, keyed by destination folder path.
    static func recordIngest(_ perFolder: [String: Int]) {
        let added = perFolder.filter { $0.value > 0 }
        guard !added.isEmpty else { return }
        mutate { s in
            for (path, n) in added {
                let key = URL(fileURLWithPath: path).standardizedFileURL.path
                let isNew = s.folders[key] == nil
                if isNew { s.shoots += 1 }
                s.folders[key, default: 0] += n
                s.photos += n
                s.addToMonth(photos: n, shoots: isNew ? 1 : 0)
            }
        }
    }

    private static func mutate(_ change: (inout AchievementState) -> Void) {
        var s = state
        change(&s)
        let unlocked = award(&s)
        save(s)
        NotificationCenter.default.post(name: didChange, object: nil)
        if !unlocked.isEmpty { NotificationCenter.default.post(name: didUnlock, object: unlocked) }
    }

    private static func award(_ s: inout AchievementState) -> [Badge] {
        let now = Date()
        var new: [Badge] = []
        for b in Badges.all where s.value(b.track) >= b.threshold && s.earned[b.id] == nil {
            s.earned[b.id] = now
            new.append(b)
        }
        return new
    }

    private static func save(_ s: AchievementState) {
        cached = s
        if let data = try? JSONEncoder().encode(s) { UserDefaults.standard.set(data, forKey: key) }
    }

    static let numberFormatter: NumberFormatter = {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        return f
    }()
}

extension Int {
    /// "2,191"
    var grouped: String { Achievements.numberFormatter.string(from: NSNumber(value: self)) ?? String(self) }
}

File: desktop/src-tauri/src/achievements.rs
//! Badges: counts photos as they come into Deadlyne and awards milestones. A photo counts when
//! it's ingested from a card, or when a folder holding it is opened for the first time; a
//! RAW+JPG pair is one photo. Per-folder high-water marks mean reopening a shoot or
//! re-ingesting a card never counts twice. Port of `Achievements.swift`, with the same badge ids.

use chrono::{Datelike, Local};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Mutex;

#[derive(Serialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Badge {
    pub id: String,
    pub track: &'static str,
    pub threshold: u64,
    pub name: &'static str,
    /// 0 bronze, 1 silver, 2 gold, 3 elite (placeholder art only).
    pub tier: u8,
}

/// Every badge, in order. Ids are permanent: earned badges and artwork files use them.
pub fn badges() -> Vec<Badge> {
    let photos: [(u64, &str); 12] = [
        (100, "Warm-Up"), (500, "Kickoff"), (1_000, "First Thousand"), (2_500, "Game Day"), (5_000, "Starter"), (10_000, "Varsity"),
        (25_000, "All-Conference"), (50_000, "All-State"), (100_000, "Hall of Fame"), (250_000, "Legend"), (500_000, "Dynasty"),
        (1_000_000, "The Million"),
    ];
    let shoots: [(u64, &str, u8); 6] = [
        (1, "First Shoot", 0), (10, "Double Digits", 0), (25, "Season Pass", 1), (50, "Road Warrior", 1), (100, "Centurion", 2),
        (250, "Iron Lens", 3),
    ];
    let mut out: Vec<Badge> = photos
        .iter()
        .enumerate()
        .map(|(i, (t, n))| Badge { id: format!("photos-{t}"), track: "photos", threshold: *t, name: n, tier: (i / 3).min(3) as u8 })
        .collect();
    out.extend(shoots.iter().map(|(t, n, tier)| Badge { id: format!("shoots-{t}"), track: "shoots", threshold: *t, name: n, tier: *tier }));
    out
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct Month {
    pub photos: u64,
    pub shoots: u64,
}

#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct State {
    pub photos: u64,
    pub shoots: u64,
    pub folders: HashMap<String, u64>,
    /// Badge id → when it was earned (Unix milliseconds).
    pub earned: HashMap<String, f64>,
    /// "2026-09" → that month's tally.
    pub months: HashMap<String, Month>,
}

static STATE: Mutex<Option<State>> = Mutex::new(None);

fn path() -> PathBuf {
    dirs::data_dir().unwrap_or_else(std::env::temp_dir).join("Deadlyne").join("achievements.json")
}

fn load() -> State {
    if let Some(s) = std::fs::read(path()).ok().and_then(|b| serde_json::from_slice(&b).ok()) {
        return s;
    }
    import_from_mac_app().unwrap_or_default()
}

/// On a Mac that ran the original Swift Deadlyne, carry its progress over once.
#[cfg(target_os = "macos")]
fn import_from_mac_app() -> Option<State> {
    let prefs = dirs::home_dir()?.join("Library/Preferences/app.deadlyne.Deadlyne.plist");
    let root = plist::Value::from_file(prefs).ok()?;
    let data = root.as_dictionary()?.get("achievements")?.as_data()?;
    let v: serde_json::Value = serde_json::from_slice(data).ok()?;
    // Swift encodes dates as seconds since 2001-01-01.
    let earned = v["earned"].as_object()?.iter().filter_map(|(k, d)| Some((k.clone(), (d.as_f64()? + 978_307_200.0) * 1000.0))).collect();
    let months = v["months"]
        .as_object()
        .map(|m| {
            m.iter()
                .map(|(k, t)| (k.clone(), Month { photos: t["photos"].as_u64().unwrap_or(0), shoots: t["shoots"].as_u64().unwrap_or(0) }))
                .collect()
        })
        .unwrap_or_default();
    let folders = v["folders"].as_object()?.iter().filter_map(|(k, n)| Some((k.clone(), n.as_u64()?))).collect();
    let s = State { photos: v["photos"].as_u64()?, shoots: v["shoots"].as_u64()?, folders, earned, months };
    save(&s);
    Some(s)
}

#[cfg(not(target_os = "macos"))]
fn import_from_mac_app() -> Option<State> {
    None
}

fn save(s: &State) {
    let p = path();
    let _ = std::fs::create_dir_all(p.parent().unwrap());
    if let Ok(bytes) = serde_json::to_vec_pretty(s) {
        let _ = crate::xmp::write_atomic(&p, &bytes);
    }
}

fn month_key() -> String {
    let now = Local::now();
    format!("{:04}-{:02}", now.year(), now.month())
}

/// Applies a change, awards any newly reached badges, saves, and returns those badges.
fn mutate(change: impl FnOnce(&mut State)) -> Vec<Badge> {
    let mut guard = STATE.lock().unwrap_or_else(|e| e.into_inner());
    let s = guard.get_or_insert_with(load);
    change(s);
    let now = Local::now().timestamp_millis() as f64;
    let mut unlocked = vec![];
    for b in badges() {
        let value = if b.track == "photos" { s.photos } else { s.shoots };
        if value >= b.threshold && !s.earned.contains_key(&b.id) {
            s.earned.insert(b.id.clone(), now);
            unlocked.push(b);
        }
    }
    save(s);
    unlocked
}

fn normalize(folder: &str) -> String {
    folder.trim_end_matches(['/', '\\']).to_string()
}

/// A folder was opened holding `count` photos; only photos beyond what was counted before are added.
pub fn record_folder(folder: &str, count: u64) -> Vec<Badge> {
    let key = normalize(folder);
    mutate(|s| {
        let old = s.folders.get(&key).copied();
        if count <= old.unwrap_or(0) {
            return;
        }
        let new_shoot = old.is_none();
        s.folders.insert(key, count);
        let added = count - old.unwrap_or(0);
        s.photos += added;
        s.shoots += new_shoot as u64;
        let m = s.months.entry(month_key()).or_default();
        m.photos += added;
        m.shoots += new_shoot as u64;
    })
}

/// Photos copied by an ingest, per destination folder.
pub fn record_ingest(per_folder: &HashMap<String, u64>) -> Vec<Badge> {
    mutate(|s| {
        for (folder, n) in per_folder.iter().filter(|(_, n)| **n > 0) {
            let key = normalize(folder);
            let new_shoot = !s.folders.contains_key(&key);
            *s.folders.entry(key).or_default() += n;
            s.photos += n;
            s.shoots += new_shoot as u64;
            let m = s.months.entry(month_key()).or_default();
            m.photos += n;
            m.shoots += new_shoot as u64;
        }
    })
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Summary {
    pub photos: u64,
    pub shoots: u64,
    pub month: Month,
    pub badges: Vec<Badge>,
    pub earned: HashMap<String, f64>,
}

#[tauri::command]
pub async fn achievements() -> Summary {
    crate::jobs::blocking(|| {
        let mut guard = STATE.lock().unwrap_or_else(|e| e.into_inner());
        let s = guard.get_or_insert_with(load);
        Summary {
            photos: s.photos,
            shoots: s.shoots,
            month: s.months.get(&month_key()).cloned().unwrap_or_default(),
            badges: badges(),
            earned: s.earned.clone(),
        }
    })
    .await
}

/// Called when a folder is opened. Returns badges unlocked by it.
#[tauri::command]
pub async fn record_folder_opened(folder: String, photos: u64) -> Vec<Badge> {
    crate::jobs::blocking(move || record_folder(&folder, photos)).await
}
